Fixes connect_loc related hard dels

ninjanomnom opened the pull request:
Probably nothing visible to players but some hard dels were happening in regards to connect_loc when their attached object was being qdeleted and not manually removing the element.

Not 100% sure this fixes them all but the dels that existed with edibles and airlocks were cleared up by these changes.

ninjanomnom committed:
Fixes connect_loc related hard dels

LemonInTheDark commented:
I see the same error on my pr #58916

LemonInTheDark commented:
I don't think `ELEMENT_DETACH` and `ELEMENT_COMPLEX_DETACH` should be allowed on the same element. One implies that you can just not delete the element on the parent's Destroy() and it'll be all good, and the other says that element.detach() REQUIRES extra context. 

It seems foolish to have both of them attached to the same thing, since `ELEMENT_DETACH` leads maintainers to believe that not calling RemoveElement is ok, when even with this change it has higher overhead attached to it, and might cause anomalous behavior.

ninjanomnom commented:
All element detach does is call detach when the object attached to is qdeleted. Complex detach doesn't imply that you *require* those additional arguments to detach and as shown here, you don't; it's just a way to specify exactly what you want detached and if no tracked object is specified then everything for that listener should be detached instead. Not calling RemoveElement *is* okay and this pr allows you to safely just qdel the object the element is attached to. I don't want to allow situations where an element needs to be detached "properly".

LemonInTheDark commented:
I can understand that, a slightly slower qdel is much more preferable then a harddelete like the ones we saw here. It's still best practice to manually remove the element tho isn't it?

LemonInTheDark commented on code/datums/elements/connect_loc.dm:
```diff
@@ -54,6 +52,15 @@
 	if (!existing && isturf(tracked.loc))
 		RegisterSignal(tracked.loc, COMSIG_TURF_CHANGE, .proc/on_turf_change)
 
+/datum/element/connect_loc/proc/unregister_all(datum/listener)
+	for(var/atom/location as anything in targets)
+		var/list/loc_targets = targets[location]
+		for(var/atom/movable/tracked as anything in loc_targets)
+			if(loc_targets[tracked] != listener)
+				continue
+			unregister_signals(listener, tracked, location)
+			UnregisterSignal(tracked, COMSIG_MOVABLE_MOVED)
```
Is this ok if there's more then one thing hooked into the listener?

ninjanomnom replied:
I was thinking about this a bit as I was working on this and I suspect there's probably a bug in here involving that, but this is how it worked before and I'd want to fix this issue separately if it is a problem.

LemonInTheDark replied:
what do you mean by worked before? 
Oh also you may want to add a if(tracked) to that unregistersignal line, since it'll runtime for the valid location of null

ninjanomnom replied:
Look in Detach, the signal was unregistered from the tracked object and didn't do any checks there either. There probably is a problem there but it'll have to be looked at in more detail in another pr

ninjanomnom committed:
Seems overriding is fine in this case

ninjanomnom committed:
Bit more aggressive cleanup

LemonInTheDark commented on code/datums/elements/connect_loc.dm:
```diff
@@ -24,25 +24,23 @@
-
-	UnregisterSignal(tracked, COMSIG_MOVABLE_LOCATION_CHANGE)
+		unregister_all(listener)
+	else if(targets[tracked.loc]) // Detach can happen multiple times due to qdel
+		unregister_signals(listener, tracked, tracked.loc)
+		UnregisterSignal(tracked, COMSIG_MOVABLE_MOVED)
```
This is dead code isn't it? since the line above it should always be run by Destroy()?

Mothblocks replied:
No, Detach can run without Destroy, and it needs to clean itself up properly.

Mothblocks replied:
Oh wait I'm dumb, you might be right.

LemonInTheDark replied:
it's dead in that nothing currently uses it, but it could be used in theory so it's not kill

LemonInTheDark commented:
![image](https://user-images.githubusercontent.com/58055496/117750992-39ba8500-b1c9-11eb-9651-675b4ee6b1ef.png)
I'm not an idiot don't worry about it please don't fire me mr headcoder

ninjanomnom committed:
It's a continue

LemonInTheDark commented:
god I give up, I'm just gonna ask ninja to reset his branch in the morning. The ref tracking stuff leaked in from my testing branch, and I can't god damn get the pr's branch to reset, no matter how much I push to it. hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh

This is the commit history it ought to be using, that's what it looked like before I got my grubby hands on it + my fixes https://github.com/LemonInTheDark/tgstation/tree/hfasd

LemonInTheDark commented:
I'm so sorry

LemonInTheDark committed:
Fixes the unit test errors, we were unregistering for the WRONG SIGNAL so we tried to remove an element that (#120)

wasn't there anymore. hhhhhhhhhhhh
Fixes the last few stack harddels, stacks were merging, and the element was being added to already qdel'd
objects, which is why they were all in null space
Adds a sanity check to AddElement to prevent this sort of thing from happening in the future
Adds ELEMENT_DETACH to the caltrop element so it passes destroy stuff along properly
